src/dist/replication/meta_server/meta_bulk_load_service.cpp

```diff
@@ -805,6 +805,46 @@ void bulk_load_service::on_partition_ingestion_reply(error_code err,
                                                      const gpid &pid)
 {
     // TODO(heyuchen): TBD
+    // if meet 2pc error, ingesting will rollback to downloading, no need to retry here
+    if (err != ERR_OK) {
+        derror_f("app({}) partition({}) ingestion files failed, error = {}", app_name, pid, err);
+        tasking::enqueue(
+            LPC_META_STATE_NORMAL,
+            _meta_svc->tracker(),
+            std::bind(&bulk_load_service::try_rollback_to_downloading, this, app_name, pid));
+        return;
+    }
+
+    if (resp.err == ERR_TRY_AGAIN && resp.rocksdb_error != 0) {
+        derror_f("app({}) partition({}) ingestion files failed while empty write, rocksdb error = "
+                 "{}, retry it later",
+                 app_name,
+                 pid,
+                 resp.rocksdb_error);
+        tasking::enqueue(LPC_BULK_LOAD_INGESTION,
+                         _meta_svc->tracker(),
+                         std::bind(&bulk_load_service::partition_ingestion, this, app_name, pid),
+                         0,
+                         std::chrono::milliseconds(10));
+        return;
+    }
+
+    // some unexpected errors happened, such as write empty write failed but rocksdb_error is ok
```

Comment:
"such as write empty write failed but rocksdb_error is ok"

Why does rocksdb allow ingesting an empty write? I have no questions on the code but just curious about this comment.

Reply:
You can see https://github.com/XiaoMi/pegasus/pull/546 firstly, which is step3, this pull request is step 4.